[migration] Update Subs, remove stale CSC in live clusters

anik120 opened the pull request:
With #198 introducing a change in a way that a datastore CatalogSource
is directly created by an OperatorSource instead of a CatalogSourceConfig,
this PR performs some upgrade logic to update/cleanup existing/stale
resources in existing clusters. Specifically:
* Subscriptions with lables "csc-owner-name" and "csc-owner-namespace" are
updated to contain information about dataStore CatalogSources
* Stale datastore CatalogSourceConfigs created before #198 are deleted.
* Stale installed CatalogSourceConfigs created during operator installation
in old clusters are deleted.

anik120 commented:
/cc @aravindhp @awgreene @galletti94

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+func deleteDatastoreCscs(cscNames []string, kubeClient client.Client) error {
+	for _, cscName := range cscNames {
```
Potential for null pointer dereference  given line 73 in `updateSubscriptions`

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+		// point the subscription to the datastore CatalogSource
+		instance.SetCatalogSource(datastoreCs.GetName())
+		instance.SetCatalogSourceNamespace("openshift-marketplace")
+		kubeClient.Update(context.TODO(), &instance)
```
Please check for error and log it.

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+	for _, cscName := range cscNames {
+		namespacedName := types.NamespacedName{Name: cscName, Namespace: "openshift-marketplace"}
+		csc := &marketplace.CatalogSourceConfig{}
+		err := kubeClient.Get(context.TODO(), namespacedName, csc)
```
Why do you need to do a `Get` here? Just attempt to delete it and log the error.

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+		return err
+	}
+	for _, csc := range cscs.Items {
+		kubeClient.Delete(context.TODO(), &csc)
```
Please check and log errors.

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+		return nil, err
+	}
+	// try and fetch the CatalogSource
+	datastoreCs := &olm.CatalogSource{}
```
Why is this a datastore `CatalogSource`?

anik120 replied:
@aravindhp this is the `CatalogSource` that a subscription will have information about after it is updated. Existing Subscriptions have information about the CSCs installed during operator installation. For example, if a CSC name `installed-community-openshift-operators` was created during installation of an operator, the Subscription had the following information
```
spec:
     source: installed-community-openshift-operators
     sourceNamespace: openshift-operators
```

This has to be changed to contain information about the datastore CatalogSource instead. 

```
spec:
     source: community-operators
     sourceNamespace: openshift-marketplace
```
Given that, shouldn't it be `datastoreCs`?

aravindhp replied:
No. Datastore `CatalogSourceConfigs` and `CatalogSources` were created by `OperatorSources` before #198.

anik120 replied:
@aravindhp what should the updated Subscription contain information about in it's `spec.Source` and `spec.SourceNamesapce` fields?

aravindhp replied:
> arent datastore CatalogSources being created directly now by OperatorSources?

Correct

> Shouldn't the updated Subscription contain information about the datastore CatalogSources in it's spec.Source and spec.SourceNamesapce fields?

Yes when you update the `Subscription`, it should reference the new datastore `CatalogSource`. The problem here was with some of the function naming that was confusing me. Given you have already pushed changes, I am unable to clarify further. In future, please wait for such conversations to get resolved before you push changes.

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+
+func deleteOldCscs(kubeClient client.Client) error {
+	options := &client.ListOptions{}
+	options.SetLabelSelector(fmt.Sprintf("opsrc-datastore: \"true\""))
```
This seems to imply that you are looking for and deleting the datastore `CatalogSourceConfigs` here but you call this function `deleteOldCscs`. This is very confusing.

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+	return cscNames, nil
+}
+
+func deleteDatastoreCscs(cscNames []string, kubeClient client.Client) error {
```
Aren't you deleting the `CatalogSourceConfigs` created as part of an operator install here? But you call the function `deleteDatastoreCscs`?

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+	return datastoreCs, nil
+}
+
+func findCsFromOpsrc(subscription *olm.Subscription, clt client.Client) (*olm.CatalogSource, error) {
```
Please use `client` instead of `clt`

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+func findCsFromOpsrc(subscription *olm.Subscription, clt client.Client) (*olm.CatalogSource, error) {
+	packageName := subscription.GetName()
```
The name of the package is in the spec of the `Subscription`

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+		return nil, err
+	}
+	for _, instance := range opsrcs.Items {
+		if isPackageInOpsrc(packageName, &instance) {
```
Code is easier to read if you check the negation here. 
```
if !isPackageInOpsrc(packageName, &instance) {
    continue
}
```
This will result in less tabbed code below.

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+			// fetch the CatalogSource with the same name
+			datastoreCs := &olm.CatalogSource{}
+			namespacedName := types.NamespacedName{Name: instance.GetName(), Namespace: "openshift-marketplace"}
+			err = clt.Get(context.TODO(), namespacedName, datastoreCs)
```
Why do you need to do a `Get` here?

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+	return false
+}
+
+func extractCsName(cscName string) (string, error) {
```
Please document sort of pattern you are looking for here.

aravindhp commented on cmd/manager/main.go:
```diff
@@ -98,6 +99,13 @@ func main() {
 	go registrySyncer.Sync(stopCh)
 	go catalogsourceconfig.Syncer.Sync(stopCh)
 
+	if os.Getenv("RELEASE_VERSION") != "" {
```
Why is this check required?

anik120 replied:
@aravindhp This is to make sure this code runs only on an OKD cluster. I noticed we set this env variable for OKD clusters and not when the operator is installed in a plain kube cluster. 
Sort of a hacky way, but the code assumes a lot of the resources are in the `openshift-marketplace` namespace. 
Although, based on your comment above about making a best effort to delete resources instead of erroring out and exiting, we could get rid of this check. The log will just look messy in a plain kube cluster.

aravindhp replied:
We should do the same steps we are performing on an OKD cluster on a plain kuber cluster. We shouldn't be making assumptions on the namespace where the operator is running. We don't do this in other parts of the code. Not sure why you think you are making that assumption here.

kevinrizza replied:
Why can't we just use a different namespace if we find it's a non openshift cluster?

aravindhp replied:
I don't want to make any assumptions on the name of the namespace.

aravindhp commented on vendor/github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1/subscription_types.go:
```diff
@@ -109,3 +109,13 @@ func NewInstallPlanReference(ref *corev1.ObjectReference) *InstallPlanReference
 }
+
+// SetCatalogSource sets the Spec.source field
+func (s *Subscription) SetCatalogSource(csName string) {
```
You cannot modify vendor code like this. This needs a PR against OLM or we carry the code.

aravindhp commented on cmd/manager/main.go:
```diff
@@ -98,6 +99,13 @@ func main() {
 
+	if os.Getenv("RELEASE_VERSION") != "" {
+		client, err := client.New(config.GetConfigOrDie(), client.Options{})
+		clean := NewCleanUpResources(log.NewEntry(log.New()), client)
```
You need to update the `ClusterOperator` CR during this process to show that we are performing some upgrade logic.

awgreene replied:
We also want to make sure that the `ClusterOperator` is not set to available while resources are being migrated - If the Upgrade path is 4.1->4.2->4.3, we don't want the upgrade from 4.2->4.3 to kickoff before this completes.

aravindhp replied:
Each cluster upgrade is a distinct operation. I don't think the 4.2 --> 4.3 option will be available until 4.1 --> 4.2 is complete.

awgreene replied:
I agree - but if marketplace reports that progression has finished we risk a race condition.

aravindhp replied:
I don't understand how there is a race condition. If marketplace has not finished migration, then the cluster is not ready for the next upgrade.

aravindhp commented:
Please add an e2e for this that asserts that no datastore `CatalogSourceConfigs` are present on the system.

anik120 commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+)
+
+// NewCleanUpResources returns a CleanUpCrew that updates existing
+// subscriptions linked to old CSCs and links them to the datastore CSCs
```
@aravindhp Is `updates existing subscriptions pointed to stale CSCs through the field spec.source and points them to datastore CatalogSources` better?

aravindhp replied:
`Subscriptions` don't point to `CatalogSourceConfigs`. They point to `CatalogSources`. This also gives rise to the question whether we delete these stale `CatalogSources` too.

anik120 replied:
So the `spec.source` field of a subscription points to a `catalogSource` which has the same name as the CSC that created it? In that case, won't just deleting the CSC delete the `CatalogSource` too as it's one of the CSC's child resource?

aravindhp replied:
>So the spec.source field of a subscription points to a catalogSource which has the same name as the CSC that created it?

This is true for the `datastore` `CatalogSourceConfigs` but I am not sure about the ones that are created as part of the old operator install flow.

>In that case, won't just deleting the CSC delete the CatalogSource too as it's one of the CSC's child resource?

Yeah I think the finalizers should handle that.

galletti94 commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+// subscriptions linked to old CSCs and links them to the datastore CSCs
+// in the openshift-marketplace namespace, and then deletes those CSCs
+// along with existing datastore CSCs.
+func NewCleanUpResources(logger *logrus.Entry, client client.Client) CleanUpCrew {
```
+1 for `Upgrader`

aravindhp commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+		}
+		// point the subscription to the datastore CatalogSource
+		instance.SetCatalogSource(datastoreCs.GetName())
+		instance.SetCatalogSourceNamespace("openshift-marketplace")
```
I see where you are hard-coding the namespace. This is not required. You should be able to infer this. This applies to all instances where you are doing this.

kevinrizza commented on cmd/manager/cleanup.go:
```diff
@@ -0,0 +1,178 @@
+package main
```
In general, I would like to state that this seems like the wrong abstraction. It is unclear what is being "cleaned" and "why" we are cleaning. I think that instead that we would be better with a migrator of some kind, if we feel like an abstraction is needed. And that migrator should know about what specific migration is happening, not that just generically we need to "clean up".

Also, have we considered if this code will stick around forever? Or is this something we will remove after openshift 4.2 release? It seems like if (more likely, when) we need to do another such migration it wouldn't make sense to do these checks as well.

anik120 replied:
@kevinrizza what if we moved all of these to a separate package called `migrator`, and document why we need this `migrator` to migrate from a openshift 4.1.1 release to a 4.1.2? (Hope I'm getting the versions correct). 
That way, when we need to do another migration, the whole abstraction can be removed.

aravindhp replied:
4.1.x to 4.2.0 is what we are concerned about here.

aravindhp replied:
I am +1 on `migrator` too.

awgreene commented on deploy/role.yaml:
```diff
@@ -22,6 +22,21 @@ rules:
   - create
   - get
   - update
+- apiGroups:
```
This change needs to be applied to `upstream` as well.

kevinrizza commented on cmd/manager/main.go:
```diff
@@ -98,6 +100,19 @@ func main() {
+	if err != nil {
+		exit(err)
+	}
+	upgrader := migrator.NewUpgrader(log.NewEntry(log.New()), client)
```
These seem like implementation details that should be obfuscated by the migrator. Why can't we just do all of this with one call? We should just do something like:

```
upgrader := migrator.NewUpgrader(log.NewEntry(log.New()), client)
err := upgrader.Migrate()
if err != nil {
	exit(err)
}
```

And you should implement a method on that migrator type that updates subscriptions and cleans up resources as implementation details.

If we start adding the details of _how_ we do the migration, especially if there are a lot of steps in future iterations, it will be really hard to read `main.go` -- which to me is already too long and complex.

kevinrizza replied:
As an aside, this will also make it easier to write unit tests for the migrator package.

kevinrizza commented on pkg/migrator/cleanup.go:
```diff
@@ -0,0 +1,246 @@
+
+// setSubsCatalogSource sets the Spec.source and Spec.sourceNamespace fields of a
+// Subscription, and returns the subscription
+func setSubsCatalogSource(subscription *olm.Subscription, csName string, csNamespace string) *olm.Subscription {
```
If you used namespaced name here, it would be simpler. Also, what is the value of this method? It seems like we only call it in one place.

kevinrizza commented on pkg/migrator/cleanup.go:
```diff
@@ -0,0 +1,246 @@
+	return subscription
+}
+
+func helperNewCatalogSourceConfig(namespace, name string) *v2.CatalogSourceConfig {
```
This function seems named poorly. We should, in general, avoid a noun like `helper` when we are just describing a function instead of a type. I would buy that a `helper` would implement a method like `GenerateCatalogSourceConfig` and be called as `helper.GenerateCatalogSourceConfig()`.  But since there is no object, we should just call it something like `func GenerateCatalogSourceConfig()`.


Also you created this `cscInfo` type which has the fields you use as function arguments, but you don't use it here.

kevinrizza commented on pkg/migrator/cleanup.go:
```diff
@@ -0,0 +1,246 @@
+// `installed-community-openshift-operators`,
+// extractCsName returns `community-operators` as output
+func extractCsName(cscName string) (string, error) {
+	var stringBuilder strings.Builder
```
strings.Builder seems like slight overkill here, no? If we are just appending two strings why not just use `fmt.Sprintf("%s-%s", possibleCsName, "-operators")`?

kevinrizza commented on pkg/migrator/cleanup.go:
```diff
@@ -0,0 +1,246 @@
+
+// getSubsPackageName returns the packageName of a Subscription
+func getSubsPackageName(subscription *olm.Subscription) string {
+	return subscription.Spec.Package
```
This function seems like overkill for a single liner that is called once and just gets the field. Why not just do it inline?

kevinrizza commented on pkg/migrator/cleanup.go:
```diff
@@ -0,0 +1,246 @@
+		if err != nil {
+			logger.Errorf("[migration] Failed to delete CatalogSourceConfig %s. Error: %s", csc.GetName(), err.Error())
+		} else {
+			logger.Printf("[migration] Succesfully deleted datastore CSC %s", csc.GetName())
```
This is technically not really true. It just means that the client returned a success and it was scheduled for deletion. We don't know that the `csc` was actually deleted yet. In fact, it won't be deleted until the finalizer is called and is executed, so there is other code in the operator that may have not been called yet. Since that's the case, I would argue that we shouldn't log this message. Instead, we could say something like "CSC %s scheduled for deletion"

kevinrizza commented on pkg/migrator/cleanup.go:
```diff
@@ -0,0 +1,246 @@
+		if err != nil {
+			logger.Errorf("[migration] Failed to delete installed CSC %s with error: ", cscInfo.name, err.Error())
+		} else {
+			logger.Printf("[migration] Succesfully deleted stale CSC %s,", cscInfo.name)
```
Same argument as other deletion message.

kevinrizza commented on pkg/migrator/cleanup.go:
```diff
@@ -0,0 +1,246 @@
+	}
+	for _, csc := range cscs.Items {
+		err = kubeClient.Delete(context.TODO(), &csc)
+		if err != nil {
```
Why don't we ever try to return this error? I understand the desire to try on every item before we give up, but we could aggregate the error and return a list of errors at the end, or even aggregate the errors into a single error when returning.

kevinrizza commented on pkg/migrator/cleanup.go:
```diff
@@ -0,0 +1,246 @@
+			logger.Printf("[migration] Succesfully deleted stale CSC %s,", cscInfo.name)
+		}
+	}
+	return nil
```
This method never actually returns an error, why always return nil? We should aggregate the kubeclient's errors and return those if they exist.

anik120 commented:
@aravindhp @kevinrizza @awgreene addressed your comments. PTAL.

aravindhp commented on test/helpers/helpers.go:
```diff
@@ -352,3 +369,135 @@ func CheckOpsrcChildResourcesDeleted(client test.FrameworkClient, opsrcName stri
+// has been succesfully updated on cluster upgrade from openshift 4.1.x to
+// openshift 4.2.0 to contain information about the datastore CatalogSource
+// instead of the installed CatalogSource
+func CheckSubsciptionSuccessfulUpdation(client test.FrameworkClient, name, namespace string) error {
```
Subscription is misspelled. Moreover is this actually an helper? This seems like the test itself.

aravindhp commented on test/helpers/helpers.go:
```diff
@@ -352,3 +369,135 @@ func CheckOpsrcChildResourcesDeleted(client test.FrameworkClient, opsrcName stri
+	if err != nil {
+		return err
+	}
+	// Check that subscription has infromation about the datastore CatalogSource,
```
Information is misspelled. Moreover please say references the datastore CatalogSource.

aravindhp commented on test/testgroups/migrationtests.go:
```diff
@@ -0,0 +1,44 @@
+	err = helpers.CreateRuntimeObject(test.Global.Client, ctx, helpers.CreateDatastoreCsc(helpers.TestDatastoreCscName, namespace))
+	require.NoError(t, err, "Could not create datastore CatalogSourceConfig")
+	// Wait for the corresponding datastore CatalogSource to deploy successfully
+	helpers.WaitForDeployment(helpers.TestDatastoreCscName, namespace, test.Global.Client)
```
You need to use `CheckCscChildResourcesCreated()` and ensure (`require.NoError`) that it succeeded.

aravindhp commented on test/testsuites/migrationtests.go:
```diff
@@ -0,0 +1,51 @@
+}
+
+// testSubscriptionsUpdated ensures that Subscriptions created during operator installation
+// in a openshift 4.1.x cluster are updated to contain information about datastore CatalogSources
```
Please use reference instead of information. Moreover, clearly stating what you are expecting in the `Subscription` will be helpful.

aravindhp commented:
- Update the commit header to `[migration] Update existing Subscriptions and remove stale resources`
- Mention why you are updating the rules if they end up being required
- Mention the Subscriptions are created by the UI
- Please be consistent regarding using `CatalogSources` in all places instead of `catalogSources` or `catalogsources`

anik120 commented on pkg/migrator/migrator.go:
```diff
@@ -0,0 +1,231 @@
+
+	installedCscs, err := UpdateSubscriptions(m.client, m.logger)
+	if err != nil {
+		return err
```
@aravindhp `UpdateSubscriptions` does a best effort update of all the Subscriptions, and only returns an error when the `client.List` operation throws an error (say for RBAC related issues).
I have tried to classify errors in this PR into two categories: 
A) Errors encountered during updating/cleaning up some resources
B) Errors encountered otherwise (say RBAC related issues etc.) 
Type `A` errors are logged and not returned at the first error encountered. 
Type `B` errors are returned as soon as a first error is encountered.

aravindhp replied:
By returning here you are not attempting to delete the stale datastore `CatalogSourceConfigs`. Which is why I am have mentioned later that `UpdateSubscription` shouldn't return an error. It should just return a list of installed `CatalogSourceConfigs` that need to be deleted. If you ran into an error and couldn't collect any installed `CatalogSourceConfigs` then the returned list would be empty.

anik120 commented on pkg/migrator/migrator.go:
```diff
@@ -0,0 +1,231 @@
+	// try and fetch the CatalogSource
+	datastoreCs := &olm.CatalogSource{}
+	namespacedName := types.NamespacedName{Name: possibleCsName, Namespace: possibleCsNamespace}
+	err := client.Get(context.TODO(), namespacedName, datastoreCs)
```
@aravindhp trying to fetch the `CatalogSource` is more for creating a back up plan in case inferring the `CatalogSource` name from the `csc-owner-name` label fails.  
If the inferred `CatalogSource` does not exist, the `migrator` then moves onto the next strategy to find the corresponding `datastore` `CatalogSource`.

aravindhp replied:
Sure but the answer in both cases should be the same. More over if you attempted the get when someone deleted the datastore `CatalogSource` and the watch which I am adding brings it back, it will still result in a `Subscription` not being updated.

anik120 replied:
@aravindhp there are two reasons for checking if the `CatalogSource` exists: 

A) If the datastore `CatalogSource` has been deleted then it shouldn't be referenced from the `Subscription`

Sure, with the watch you're adding it'll be brought back, but there's another scenario where:

B) The infered `CatalogSource` name is wrong. For example if the value of `csc-owner-label` is set to `installed-foo-openshift-marketplace`, and the `foo-operators` `CatalogSource` does not exist simply because `foo` in `installed-foo-openshift-marketplace` was misplaced in the first place, then the `Subscription` shouldn't reference a non-existent `CatalogSource`. Checking if the `CatalogSource` exists provides us the safety net.

aravindhp replied:
Good point regarding B.

aravindhp commented on cmd/manager/main.go:
```diff
@@ -92,6 +94,18 @@ func main() {
+	// Perform migration logic to upgrade cluster from 4.1.x to 4.2.0
+	client, err := client.New(config.GetConfigOrDie(), client.Options{})
+	if err != nil {
+		exit(err)
```
Do we really want to terminate the operator if we cannot migrate? Even if the CVO restarts the operator, we could end up in the same situation again. This will basically result in a cluster that doesn't have a marketplace operator. The same comment applies for the `Migrate()` call.

kevinrizza replied:
It seems like the right thing to do would be to retry until it succeeds, given the normal pattern of eventual consistency.

aravindhp replied:
What if it is stuck here for whatever reason? We will be migrating for ever. So either we do this in a background goroutine or do a best effort migration.

anik120 replied:
> What if it is stuck here for whatever reason? 

Exactly my concern. Do we want to keep consuming resources in an effort to migrate, if we are unable to migrate fully?

anik120 replied:
Also, the Subscriptions are updated by the migrator, so unless there's a switch which indicates if a Subscription has been updated or not, all the Subscriptions will keep getting updated, albeit just one field.

Keeping track of which Subscriptions were updated means consuming even more resources.

aravindhp replied:
Then we should be removing the labels on the Subscriptions that we update

kevinrizza replied:
I think that doing it every time we start is just part of the deal. I'm not sure how else we would handle it.

Are we okay with our migration failing and then moving on? Perhaps it depends on our definition of `failure`?

anik120 replied:
> Then we should be removing the labels on the Subscriptions that we update

@aravindhp Are we sure we can do that without producing any side effects? 

> Are we okay with our migration failing and then moving on? Perhaps it depends on our definition of failure?

Maybe clean up/ update as many resources possible, and report resources that could not be cleaned/updated for admin to act on them?

aravindhp replied:
>Are we sure we can do that without producing any side effects?

Yes. Note that this does not mean removing *all* labels. Just the ones we care about.

>Maybe clean up/ update as many resources possible,

This is what we should be doing in general.

> report resources that could not be cleaned/updated for admin to act on them?

How will you report? Asking admins to look at logs is not acceptable. Please think of worst case scenarios where not updating a Subscription or not being able to delete the datastore/installed `CatalogSourceConfigs` will cause issues.

anik120 replied:
> Please think of worst case scenarios where not updating a Subscription or not being able to delete the datastore/installed CatalogSourceConfigs will cause issues.

Does that mean we're agreeing on creating a go routine that keeps making best effort to clean up resources until there's no errors?

aravindhp replied:
You shouldn't jump  to that conclusion until you explore the scenarios.

anik120 commented on pkg/migrator/migrator.go:
```diff
@@ -0,0 +1,231 @@
+	return nil
+}
+
+// UpdateSubscriptions updates the existing Subscriptions'
```
@aravindhp I was going for the plural `Subscriptions` as `updateSubscriptions` are updating multiple subscriptions.

aravindhp replied:
OK. But please update the doc to say Subscriptions reference installed `CatalogSources`. Also use `Subscriptions` everywhere instead of `subscriptions` in some places.

anik120 commented on pkg/migrator/migrator.go:
```diff
@@ -0,0 +1,231 @@
+
+	subscriptions := &olm.SubscriptionList{}
+	// Get the list of existing subscriptions that have the label "csc-owner-name"
+	err := kubeClient.List(context.TODO(), options, subscriptions)
```
> How does the "csc-owner-name" label end up in the Subscription object? Does OLM pass this along based on the referenced CatalogSource in the Subscription?

@aravindhp I remember the UI code adding the labels [here] (https://sourcegraph.com/github.com/openshift/console/-/blob/frontend/public/components/operator-hub/operator-hub-subscribe.tsx#L110), but looks like that's changed now. 

> You are also listing Subscriptions in all namespaces. Please confirm that this is required. I don't remember off the top of my head where the Subscriptions get created by the UI.

The UI creates the Subscriptions in the `openshift-operator` namespace

aravindhp replied:
> I remember the UI code adding the labels here, but looks like that's changed now.

You should be looking at the 4.1 branch

>The UI creates the Subscriptions in the `openshift-operator` namespace

Then you do need the cluster role. It basically means that the `Subscriptions` could be anywhere as you are allowed to pick a namespace for an operator and the UI will create the `OperatorGroup` for that namespace if needed. In that case I would think the `Subscription` would be in the selected namespace.

anik120 commented on pkg/migrator/migrator.go:
```diff
@@ -0,0 +1,231 @@
+		return nil, err
+	}
+	for _, instance := range subscriptions.Items {
+		newCscInfo := types.NamespacedName{Name: instance.GetLabels()[builders.OwnerNameLabel], Namespace: instance.GetLabels()[builders.OwnerNamespaceLabel]}
```
> What if csc-owner-namespace is not present?

I don't see a scenario where `csc-owner-namespace` is not present, if `csc-owner-namespace` is present. That would be an inconsistency that'd need to be addressed wherever the `Subscription` object is being created 

> PS: I don't buy the readability argument that many people give for this. Add comments instead. ;-)

I guess you foresaw my argument :)

aravindhp replied:
>I don't see a scenario where csc-owner-namespace is not present, if csc-owner-namespace is present. That would be an inconsistency that'd need to be addressed wherever the Subscription object is being created

There could be a potential scenario where an user or some other entity removes the label. You have to account for that.

anik120 commented on pkg/migrator/migrator.go:
```diff
@@ -0,0 +1,231 @@
+
+	cscs := &v2.CatalogSourceConfigList{}
+	// Get the list of existing cscs that have the lable "opsrc-datastore: "true" "
+	err := kubeClient.List(context.TODO(), options, cscs)
```
@aravindhp Problem is determining the namespace of the `marketplace-operator`. I don't suppose there's a clean way of determining the namespace of the operator (`openshift-marketplace` or `marketplace` etc). Listing the `CatalogSourceConfigs` in all namespaces seems like the cleanest way to do it.

aravindhp replied:
https://github.com/operator-framework/operator-marketplace/blob/master/cmd/manager/main.go#L49

kevinrizza commented on test/helpers/helpers.go:
```diff
@@ -352,3 +369,135 @@ func CheckOpsrcChildResourcesDeleted(client test.FrameworkClient, opsrcName stri
+// CreateSubscription returns a newly built Subscription with the lables
+// `csc-owner-name` and `csc-owner-namespace`
+func CreateSubscription(name, namespace string) *olm.Subscription {
+	lables := make(map[string]string)
```
this should be `labels`

anik120 commented on pkg/migrator/migrator.go:
```diff
@@ -0,0 +1,231 @@
+func setSubsCatalogSource(subscription *olm.Subscription, csName string, csNamespace string) *olm.Subscription {
+	subscription.Spec.CatalogSource = csName
+	subscription.Spec.CatalogSourceNamespace = csNamespace
+	return subscription
```
@aravindhp We want the subscription to be updated to the new values. 
Does that answer your question or am I misunderstanding the question?

aravindhp replied:
What I am trying to say is `updatedSub := setSubsCatalogSource(&instance, datastoreCs.GetName(), datastoreCs.GetNamespace())`  is equivalent to `setSubsCatalogSource(&instance, datastoreCs.GetName(), datastoreCs.GetNamespace())`. You can use `instance` instead of `updatedSub` given you are using a pointer. 

You might also want to consider if this function is really required given you are only calling it in one place and the amount of work that is being done in it.

anik120 replied:
Ah, true. It made sense when I had accidentally modified the vendor code, but now that I'm not doing that the function doesn't make that much sense. I'll update the object at the call site.

anik120 commented:
@aravindhp @kevinrizza addressed your comments. PTAL.
Pending discussion: https://github.com/operator-framework/operator-marketplace/pull/207#discussion_r297678569

aravindhp commented:
Continued in #223
